File: 0152-maximum-product-subarray/0152-maximum-product-subarray.cpp
class Solution {
public:
    int maxProduct(vector<int>& nums) {
        int len = nums.size();
        if (len == 0) return 0;

        int maximum = nums[0], minimum = nums[0];
        int answer = maximum;

        for (int i = 1; i < len; i++) {
            const int cur = nums[i];

            int tempMaximum = max(cur, max(cur * maximum, cur * minimum));
            minimum = min(cur, min(cur * maximum, cur * minimum));
            maximum = tempMaximum;

            answer = max(answer, maximum);
        }

        return answer;
    }
};

File: 0152-maximum-product-subarray/0152-maximum-product-subarray.java
class Solution {
    public int maxProduct(int[] nums) {
        int minp = nums[0];
        int maxp = nums[0];
        int answer = nums[0];

        for (int i = 1; i < nums.length; i++) {
            int num = nums[i];
            int currentMinp = Math.min(num, Math.min(minp*num, maxp*num));
            int currentMaxp = Math.max(num, Math.max(minp*num, maxp*num));

            minp = currentMinp;
            maxp = currentMaxp;

            answer = Math.max(answer, maxp);
        }

        return answer;
    }
}